File: 2023/day01/zig/src/main.zig
const std = @import("std");

pub fn part_1(filename: []const u8) !u32 {
    var sum: u32 = 0;
    var file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    var in_stream = buf_reader.reader();
    
    var buf: [1024]u8 = undefined;
    while (try in_stream.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        var first: u8 = 0x30;
        var last: u8 = 0x30;
        var first_found: bool = false;
        for (line) |char| {
            if ((char < 0x30) or (char > 0x39)) continue;
            if (!first_found) {
                first = char;
                first_found = true;
            }
            last = char;
        }
        sum += (first - 0x30) * 10;
        sum += last - 0x30;
    }

    return sum;
}

pub fn check_written_digit(word: []const u8, remaining_length: usize) u8 {
    if (remaining_length < 3) return 0;
    if (std.mem.eql(u8, word[0..3], "one")) return 0x31;
    if (std.mem.eql(u8, word[0..3], "two")) return 0x32;
    if (std.mem.eql(u8, word[0..3], "six")) return 0x36;
    if (remaining_length < 4) return 0;
    if (std.mem.eql(u8, word[0..4], "four")) return 0x34;
    if (std.mem.eql(u8, word[0..4], "five")) return 0x35;
    if (std.mem.eql(u8, word[0..4], "nine")) return 0x39;
    if (remaining_length < 5) return 0;
    if (std.mem.eql(u8, word[0..5], "three")) return 0x33;
    if (std.mem.eql(u8, word[0..5], "seven")) return 0x37;
    if (std.mem.eql(u8, word[0..5], "eight")) return 0x38;
    return 0;
}

pub fn part_2(filename: []const u8) !u32 {
    var sum: u32 = 0;
    var file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    var in_stream = buf_reader.reader();
    
    var buf: [1024]u8 = undefined;
    while (try in_stream.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        var first: u8 = 0x30;
        var last: u8 = 0x30;
        var first_found: bool = false;
        var idx: u32 = 0;
        while (idx < line.len) : (idx += 1) {
            var char: u8 = line[idx];
            if ((char < 0x30) or (char > 0x39)) {
                char = check_written_digit(line[idx..line.len], line.len - idx);
                if (char == 0)
                {
                    continue;
                }
            }
            if (!first_found) {
                first = char;
                first_found = true;
            }
            last = char;
        }
        sum += (first - 0x30) * 10;
        sum += last - 0x30;
    }

    return sum;
}

pub fn main() !void {

    var args = std.process.args();
    _ = args.skip();

    const stdout_file = std.io.getStdOut().writer();
    var bw = std.io.bufferedWriter(stdout_file);
    const stdout = bw.writer();

    const filename = args.next() orelse {
        std.debug.print("Usage: <prog> <filename>\n", .{});
        std.os.exit(1);
    };
    
    const part_1_result: u32 = try part_1(filename);
    try stdout.print("Part 1: {}\n", .{part_1_result});
    
    const part_2_result: u32 = try part_2(filename);
    try stdout.print("Part 2: {}\n", .{part_2_result});

    try bw.flush(); // don't forget to flush!
}

test "Part 1" {
    const expected: u32 = 142;
    const filename: []const u8 = "../test.txt";
    const actual: u32 = try part_1(filename);
    try std.testing.expectEqual(expected, actual);
}

test "Part 2" {
    const expected: u32 = 281;
    const filename: []const u8 = "../test_2.txt";
    const actual: u32 = try part_2(filename);
    try std.testing.expectEqual(expected, actual);
}

File: 2023/day01/rust/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::env;

fn str_to_vec(str: String) -> Vec<char> {
    let mut v: Vec<char> = vec![];
    for ch in str.chars() {
        v.push(ch);
    }
    return v;
}

fn part_1(input: Vec<String>) -> u32 {
    let mut sum: u32 = 0;
    
    for line in input {
        let chars: Vec<char> = str_to_vec(line);
        let mut first: u32 = 0x30;
        let mut last: u32 = 0x30;
        let mut first_found: bool = false;
        for ch in chars {
            if (ch as u8) < 0x30 || (ch as u8) > 0x39 {
                continue;
            }
            if !first_found {
                first = ch as u32;
                first_found = true;
            }
            last = ch as u32;
        }
        sum += (first - 0x30) * 10;
        sum += last - 0x30;
    }

    return sum;
}

fn check_written_digit(word: &str, len: usize) -> u8 {
    if len < 3 { return 0; }
    if &word[0..=2] == "one" { return 0x31; }
    if &word[0..=2] == "two" { return 0x32; }
    if &word[0..=2] == "six" { return 0x36; }
    if len < 4 { return 0; }
    if &word[0..=3] == "four" { return 0x34; }
    if &word[0..=3] == "five" { return 0x35; }
    if &word[0..=3] == "nine" { return 0x39; }
    if len < 5 { return 0; }
    if &word[0..=4] == "three" { return 0x33; }
    if &word[0..=4] == "seven" { return 0x37; }
    if &word[0..=4] == "eight" { return 0x38; }
    return 0;
}

fn part_2(input: Vec<String>) -> u32 {
    let mut sum: u32 = 0;
    
    for line in input {
        let chars: Vec<char> = str_to_vec(line.clone());
        let mut first: u32 = 0x30;
        let mut last: u32 = 0x30;
        let mut first_found: bool = false;
        for idx in 0..chars.len() {
            let mut ch: u8 = chars[idx] as u8;
            if (ch as u8) < 0x30 || (ch as u8) > 0x39 {
                ch = check_written_digit(&line[idx..chars.len()], chars.len() - idx);
                if ch == 0 {
                    continue;
                }
            }
            if !first_found {
                first = ch as u32;
                first_found = true;
            }
            last = ch as u32;
        }
        sum += (first - 0x30) * 10;
        sum += last - 0x30;
    }

    return sum;
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    return Ok(io::BufReader::new(file).lines());
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        println!("Usage: <prog> <filename>");
        return;
    }
    let filename: &str = &args[1];
    if let Ok(lines) = read_lines(filename) {
        let mut input: Vec<String> = vec![];
        for line in lines {
            if let Ok(e) = line {
                input.push(e);
            }
        }

        println!("Part 1: {}", part_1(input.clone()));
        println!("Part 2: {}", part_2(input));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn part_1_test() {
        if let Ok(lines) = read_lines("../test.txt") {
            let mut input: Vec<String> = vec![];
            for line in lines {
                if let Ok(e) = line {
                    input.push(e);
                }
            }
            let res = part_1(input);
            assert_eq!(142, res);
        }
    }
    #[test]
    fn part_2_test() {
        if let Ok(lines) = read_lines("../test_2.txt") {
            let mut input: Vec<String> = vec![];
            for line in lines {
                if let Ok(e) = line {
                    input.push(e);
                }
            }
            let res = part_2(input);
            assert_eq!(281, res);
        }
    }
}
